### PR TITLE
Add integration testing, and fix exit code issues with 2.22.0

### PR DESCRIPTION
**Pull request by ukd1**

### CODE REVIEW

**Comment by ukd1** [comment]
@jbarber @magni- @maciejgryka this adds tests (outside of go, just to make sure), plus fixes the exit code issue I think. It should also stop it happening again, as these tests are required to release now.

* https://app.circleci.com/pipelines/github/rainforestapp/rainforest-cli/214/workflows/731f3316-9787-4f90-bd94-2ae1c804ad7e - failing, showing the issue in the code before b4581207fea74a70fdcf9151233e47fdc6492339
* https://app.circleci.com/pipelines/github/rainforestapp/rainforest-cli/215/workflows/7dd00687-4b44-4689-be45-181ac68a3d09 - passing at 00883034a0a4f86ad236358bf7eaf525a588a759, which switches to spawning instead of replacing the current process.

Thoughts? Any more tests I could add easily?

**Review by jbarber — COMMENTED** [review]
Good idea adding the tests.

I have problems with the wider context of this implementation:
* I don't understand why we're spawning a new version of the CLI to get the report, can't we call `writeJunit` directly from `monitorRunStatus`? This would save us messing about having to spawn a new instance of the CLI which is always going to be error prone because doing things with other processes is always complex and difficult to get right

I would strongly advise we take this opportunity to change the current approach of spawning a new CLI instance to calling the internal functions for getting the jUnit report directly.

Other things that occurred to me on reviewing this:
* `postRunJUnitReport` is called from `monitorRunStatus` after the run is complete and before deciding to invoke rerun. I don't think this is correct because it could result a junit file to be written for the wrong run (e.g. run fails -> report is written -> rerun is invoked -> CLI crashes (run passes) -> user gets the wrong junit output)
* Using `--token` on the command line is insecure because it exposes the user's token to inspection via `ps` / etc. It should be communicated to the child process via env vars which aren't visible to other users unless you have special permissions
* At the moment when we've started a run and we fail to retrieve the junit report we don't do anything, which is not ideal. Failing the CI/CD run because of this doesn't seem like a good idea, but maybe we could fallback to emitting a minimal XML file with the passed/failed status in it?

File: resources.go [hunk 1]
```diff
@@ -175,9 +176,10 @@ func postRunJUnitReport(c cliContext, runID int) error {
 		return cli.NewExitError(err.Error(), 1)
 	}
 
-	exec_err := syscall.Exec(path, cmd, os.Environ())
-	if exec_err != nil {
-		return cli.NewExitError(exec_err.Error(), 1)
+	rfCmd := exec.Command(path, strings.Join(cmd, " "))
+	_, err = rfCmd.Output()
+	if err != nil {
+		return cli.NewExitError(err.Error(), 1)
```
[the remark this answers is not in the document]
Comment (ukd1) [reply resources.go:182]: That's my understanding too.

**Comment by ukd1** [comment]
> Good idea adding the tests.

Ya, apparently there weren't any for exit status / monitor run

> I have problems with the wider context of this implementation:
> 
> * I don't understand why we're spawning a new version of the CLI to get the report, can't we call `writeJunit` directly from `monitorRunStatus`? This would save us messing about having to spawn a new instance of the CLI which is always going to be error prone because doing things with other processes is always complex and difficult to get right
> 
> I would strongly advise we take this opportunity to change the current approach of spawning a new CLI instance to calling the internal functions for getting the jUnit report directly.

Well, ya I agree - but this isn't really changing it, yet that's also where the problem came from - I re-used the code from the re-run functionallity, but also didn't understand that it _replaced_ the process, rather than did what this code does now.

> Other things that occurred to me on reviewing this:
> 
> * `postRunJUnitReport` is called from `monitorRunStatus` after the run is complete and before deciding to invoke rerun. I don't think this is correct because it could result a junit file to be written for the wrong run (e.g. run fails -> report is written -> rerun is invoked -> CLI crashes (run passes) -> user gets the wrong junit output)

Yes, but:
1. this is how it worked before
2. even though that made no-sense to me - who wants multiple junit files, appended with numbers - but see 1

> * Using `--token` on the command line is insecure because it exposes the user's token to inspection via `ps` / etc. It should be communicated to the child process via env vars which aren't visible to other users unless you have special permissions

Correct, and great point - though if you notice, it only does it if it's already done in the first place - rather than change their env. So, I think this is actually cleaner, and just as risky as they're making the command already.

> * At the moment when we've started a run and we fail to retrieve the junit report we don't do anything, which is not ideal. Failing the CI/CD run because of this doesn't seem like a good idea, but maybe we could fallback to emitting a minimal XML file with the passed/failed status in it?

You mean we'll just hang if there is no report returned?


:+1: re moving to calling the internal functions for getting the jUnit report directly. That should remove a number of these issues.

File: resources.go [hunk 2]
```diff
@@ -175,9 +176,10 @@ func postRunJUnitReport(c cliContext, runID int) error {
 		return cli.NewExitError(err.Error(), 1)
 	}
 
-	exec_err := syscall.Exec(path, cmd, os.Environ())
-	if exec_err != nil {
-		return cli.NewExitError(exec_err.Error(), 1)
+	rfCmd := exec.Command(path, strings.Join(cmd, " "))
```
[the remark this answers is not in the document]
Comment (ukd1) [reply resources.go:179]: 👀

**Comment by jbarber** [comment]
> > Good idea adding the tests.
> 
> Ya, apparently there weren't any for exit status / monitor run
> 
> > I have problems with the wider context of this implementation:
> > 
> > * I don't understand why we're spawning a new version of the CLI to get the report, can't we call `writeJunit` directly from `monitorRunStatus`? This would save us messing about having to spawn a new instance of the CLI which is always going to be error prone because doing things with other processes is always complex and difficult to get right
> > 
> > I would strongly advise we take this opportunity to change the current approach of spawning a new CLI instance to calling the internal functions for getting the jUnit report directly.
> 
> Well, ya I agree - but this isn't really changing it, yet that's also where the problem came from - I re-used the code from the re-run functionallity, but also didn't understand that it _replaced_ the process, rather than did what this code does now.
> 
> > Other things that occurred to me on reviewing this:
> > 
> > * `postRunJUnitReport` is called from `monitorRunStatus` after the run is complete and before deciding to invoke rerun. I don't think this is correct because it could result a junit file to be written for the wrong run (e.g. run fails -> report is written -> rerun is invoked -> CLI crashes (run passes) -> user gets the wrong junit output)
> 
> Yes, but:
> 
> 1. this is how it worked before
> 2. even though that made no-sense to me - who wants multiple junit files, appended with numbers - but see 1

Ah, I didn't realize/forgot we did that.
 
> > * Using `--token` on the command line is insecure because it exposes the user's token to inspection via `ps` / etc. It should be communicated to the child process via env vars which aren't visible to other users unless you have special permissions
> 
> Correct, and great point - though if you notice, it only does it if it's already done in the first place - rather than change their env. So, I think this is actually cleaner, and just as risky as they're making the command already.

I did not notice!
 
> > * At the moment when we've started a run and we fail to retrieve the junit report we don't do anything, which is not ideal. Failing the CI/CD run because of this doesn't seem like a good idea, but maybe we could fallback to emitting a minimal XML file with the passed/failed status in it?
> 
> You mean we'll just hang if there is no report returned?

I was thinking more that ATM the endpoint could fail to return the junit (junit endpoint times out/whatever) and the CLI will exit without any XML being produced - perhaps we should notice there's no XML and generate something? I don't know if this a good idea or not.

**Comment by ukd1** [comment]
/reviewme @jbarber

🙇🏻 

...obvi, will squash a little

**Review by jbarber — COMMENTED** [review]
Looks good!

Bunch of suggestions about refactoring the YAML - none of them are critical though.

Additional thoughts I had and will leave you to ponder - totally fine to leave for another PR IMO:
* Is it worth making the integration tests dependent on the `test` job - so we don't run them unless the unit tests pass (or maybe just run them on release?), as I'm not sure it's worth running them all on all platforms for all commits because of cost (not sure what this is though!) and potential sources of flakiness. Although I can see the benefits of doing them all the time as well
* The integration tests aren't testing the artifacts we're going to be shipping that were built on a Linux environment - so there's some potential that the thing we ship doesn't work because of differences between the go lang compiler on linux vs OSX vs Windows. I don't think it's a major hole in coverage and something we can ignore for now because I think copying artifacts between CircleCI jobs will be a PITA
* The integration tests aren't testing the rerun functionality, which would be a nice to have? Test would be to execute the failing run group with rerun set to 1 and check the appropriate junit files are created?

**Review by jbarber — APPROVED** [review]

File: .circleci/config.yml [hunk 3]
```diff
@@ -12,6 +22,241 @@ jobs:
...
+          name: Install go
+          command: |
+            curl -L https://golang.org/dl/go1.17.1.darwin-amd64.tar.gz > go.tar.gz
+            tar -xzf go.tar.gz
+            rm go.tar.gz
+            ./go/bin/go version
+      - run:
+          name: Build
+          command: |
+            ./go/bin/go build
```
[the remark this answers is not in the document]
Comment (ukd1) [reply .circleci/config.yml:39]: ya, but I liked it to all look consistent, even if it could be one line less :)

File: .circleci/config.yml [hunk 4]
```diff
@@ -12,6 +22,241 @@ jobs:
...
+      - checkout
+      - run:
+          name: "Build, and run rainforest"
+          shell: powershell.exe
+          command: |
+            go build
+            .\rainforest-cli.exe -v --skip-update
+            .\rainforest-cli.exe --skip-update run --run-group 9503
+            if ($LastExitCode -gt 0) {
+              echo "TESTING ::: Got the expected non-zero exit code."
```
Comment (magni-) [inline .circleci/config.yml:233]: Should we expect 1 specifically?

File: .circleci/config.yml [hunk 5]
```diff
@@ -12,6 +22,241 @@ jobs:
...
+              # exit 0
```
Comment (magni-) [inline .circleci/config.yml:234]: These exit codes shouldn't be commented out, right?

File: .circleci/config.yml [hunk 6]
```diff
@@ -31,10 +276,34 @@ workflows:
   version: 2
   test_and_deploy:
     jobs:
-      - test:
-          context:
-            - DockerHub
+      - test
+      - integration_mac_pass
```
Comment (magni-) [inline .circleci/config.yml:280]: You can use a [matrix](https://circleci.com/docs/2.0/configuration-reference/#matrix-requires-version-21) for this:
```suggestion
      - integration_pass:
          matrix:
            parameters:
              os: [linux, macos, windows]
```

And then you'd define a parametrized `integration_pass` job (not sure this is valid YAML, I'm just typing this out on GitHub)
```yaml
jobs:
  - integration_pass:
    parameters:
      os:
        type: executor
     steps:
       - checkout
       - ./$RAINFOREST_CLI -v --skip-update # set the env var to rainforest-cli or rainforest-cli.exe at the executor level
       - run:
           name: Create passing Rainforest run
           command: ./$RAINFOREST_CLI --skip-update run --run-group 9503
```

File: .circleci/config.yml (hunk 4 above)
Comment (ukd1) [reply .circleci/config.yml:233]: great point, this should stop it "working" if we change the cli commands

File: .circleci/config.yml [hunk 7]
```diff
@@ -12,6 +22,241 @@ jobs:
...
+            ./go/bin/go version
+      - run:
+          name: Build
+          command: |
+            ./go/bin/go build
+      - run:
+          name: Start a run that will pass, with junit
+          command: |
+            ./rainforest-cli -v --skip-update
+            ./rainforest-cli --skip-update run --run-group 9502 --junit-file junit-pass.xml
```
[the remark this answers is not in the document]
Comment (ukd1) [reply .circleci/config.yml:61]: done!

**Pull request merged**